docs/structured-streaming-kafka-integration.md

```diff
@@ -457,8 +463,17 @@ The following configurations are optional:
   <td>string</td>
   <td>none</td>
   <td>streaming and batch</td>
+  <td>Sets the topic that all rows will be written to in Kafka. This option overrides
+  ```path``` option and any topic column that may exist in the data.</td>
+</tr>
+<tr>
+  <td>path</td>
+  <td>string</td>
+  <td>none</td>
+  <td>streaming and batch</td>
   <td>Sets the topic that all rows will be written to in Kafka. This option overrides any
-  topic column that may exist in the data.</td>
+  topic column that may exist in the data and is overridden by ```topic``` option.
```

Comment:
How about the use-case when you want to have 'topic' column in your data with different topic values in it (ex: "win", "click","visit"), but want to write that data to a single topic, so you override topic setting by topic option?

Reply:
Not yet understand the use-case, can you show the smallest example?

Reply:
I rethought the situation and understood that you have the point, such situation is impossible, as we can only have "key", "value" and "topic" columns in the DataFrame that is written to Kafka.

Reply:
I agree with you, that all three of them should be checked. However, now we have the validation being  splitted and duplicated between KafkaSourceProvider#validateQuery, KafkaWriteTask#createProjection and KafkaSourceProvider#resolveTopic. We can refactor this moment and move this validations to one place, or just leave as is and add topic column and topic/path option comparison validation to KafkaSourceProvider#validateQuery. The fist is more complicated and error-prone, but will result in more readable code. On the other hand, second solution will not require so much code to be rewritten. Which way do you think is right?

Reply:
I agree the first solution is more complicated but not fully understand why is it error-prone? Having checks in 2 different places which are belonging together is the error-prone solution. Of course good unit test coverage is required.

Reply:
@gaborgsomogyi I agree with you, that first solution seems to be right way to go. I hope that I will not miss any place where it is used, while writing unit tests. My concern is about checking the topic column value. We cannot move the projection out of KafkaWriteTask, but if we move all the validation inside, than it would no longer be fail-fast: in current implementation we are checking all that we can before RDD creation. Although, I do not really understand why checks are duplicated between [KafkaWriteTask#createProjection](https://github.com/apache/spark/blob/master/external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaWriteTask.scala#L101) and [KafkaWriter#validateQuery](https://github.com/apache/spark/blob/master/external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaWriter.scala#L45)? As I can see validateQuery is always called before createProjection. @gaborgsomogyi @jaceklaskowski do you know any reason for this duplication?
![image](https://user-images.githubusercontent.com/10544767/54438390-a059ca80-4747-11e9-8416-13a16009e2bd.png)

The solution I propose: 
- remove duplicated checks from KafkaWriteTask#createProjection
- move check for topic and path option match to KafkaWriter#validateQuery
- add topic option and topic attribute match validation to KafkaWriteTask

Reply:
KafkaWriteTask and KafkaWriter are not public interfaces, and in all places where [KafkaRowWriter#createProjection](https://github.com/apache/spark/blob/master/external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaWriteTask.scala#L101) is called, it is preceded by [KafkaWriter#validateQuery](https://github.com/apache/spark/blob/master/external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaWriter.scala#L45). I do not understand the purpose of the double check between these methods? It was introduced in #17043, @tcondie @tdas could you please explain the purpose of validation duplication? 

![image](https://user-images.githubusercontent.com/10544767/54684731-27cb8300-4b26-11e9-80de-d6c7bf2f3755.png)
@gaborgsomogyi the reason, why first solution is error-prone, is that I don't actually understand the reason of the decisions taken. When I removed the validation from `KafkaRowWriter#createProjection` nothing failed.

And the second concern is as follows: in the topic column we can specify different values and our messages will be sent to different topics based on the value in particular row. When the topic option is specified it is applied for all messages, even if the topic column is present and has different values. So, I think that these two options have different purposes of use and do not intersect, so checking them for equality is just redundant. I can imagine only one situation when I will specify both column and option: when I want to use column approach with different values in production, but for testing purposes I specify topic option so that all messages come to one topic for a while.

And I think that is better to leave the [KafkaSourceProvider#resolveTopic](https://github.com/apache/spark/pull/23791/files#diff-eeac5bdf3a1ecd7b9f8aaf10fff37f05R197) function where it is: 
we should not only validate that topic path and option match when both are defined but also trim and decide on which value to take (as one of them or both may be null). If we move method to `KafkaWriter` we will have to call this method from several other classes and pass additional path param and now all that logic is incapsulated inside `KafkaSourceProvider`.

I introduced the approach described here in my last commit.

Reply:
I've taken a deeper look and as you've mentioned it's a valid use-case where different topics are written because the topic column value differs. As a result the originally added fallback solution can be the way.